File: 2024_fase1/solutions/molecularscissors.c
#include <stdio.h>
#include <string.h>

#define MIN(a, b) ((a < b)? a: b)

char complement(char c){
  switch(c){
    case 'A':
      return 'T';
    case 'T':
      return 'A';
    case 'C':
      return 'G';
    case 'G':
      return 'C';
  }
}

void solve(char s[]){
  size_t n = strlen(s);
  int radii[100001];
  for(int i=0; i<n; i++)
    radii[i] = 0;

  int left = 0;
  int right = -1;
  int best_start = -1;
  int best_len = 0;

  for(int center=0; center<n; center++){
    int radius = 0;
    if(center <= right){
      int mirror = left + right - center + 1;
      radius = MIN(radii[mirror], right-center+1);
    }

    while(center - radius - 1 >= 0
      && center + radius < n
      && s[center - radius - 1] == complement(s[center + radius])
    ){
      radius++;
    }

    radii[center] = radius;

    if(center + radius - 1 > right){
      left = center - radius;
      right = center + radius -1;
    }

    int length = 2*radius;
    if(length >= 4 && length > best_len){
      best_start = center - radius;
      best_len = length;
    }
  }

  if(best_len == 0)
    printf("false\n");
  else
    printf("%d %d\n", best_start+1, best_len);
}

int main(){
  char buff[100001];

  while(scanf("%s", buff) != EOF)
    solve(buff);

  return 0;
}

File: 2024_fase1/solutions/molecularscissors.py
import sys

COMPLEMENT = {
  'A': 'T',
  'T': 'A',
  'C': 'G',
  'G': 'C',
}

def longest_site(dna):
  n = len(dna)
  radii = [0] * n
  left = 0
  right = -1
  best_start = -1
  best_len = 0

  for center in range(n):
    if center > right:
      radius = 0
    else:
      mirror = left + right - center + 1
      radius = min(radii[mirror], right - center + 1)

    while (
      center - radius - 1 >= 0
      and center + radius < n
      and dna[center - radius - 1] == COMPLEMENT[dna[center + radius]]
    ):
      radius += 1

    radii[center] = radius

    if center + radius - 1 > right:
      left = center - radius
      right = center + radius - 1

    length = 2 * radius
    if length >= 4 and length > best_len:
      best_start = center - radius
      best_len = length

  if best_len == 0:
    return None

  return best_start + 1, best_len

for line in sys.stdin:
  dna = line.strip()
  answer = longest_site(dna)
  if answer is None:
    print('false')
  else:
    print(answer[0], answer[1])
